addons/delivery/models/stock_picking.py

```diff
@@ -192,6 +192,18 @@ def _add_delivery_cost_to_so(self):
         sale_order = self.sale_id
         if sale_order.invoice_shipping_on_delivery:
             sale_order._create_delivery_line(self.carrier_id, self.carrier_price)
+        else:
```

Comment:
add a comment:
If the delivery line is of real cost type, reset its name and set its `price_unit` to the real cost.